### PR TITLE
refactor: migrate cluster-related code to TypeScript

### PR DESCRIPTION
**Pull request by luin**

### CODE REVIEW

File: lib/cluster/index.ts [hunk 1]
```diff
@@ -0,0 +1,663 @@
...
+var asCallback = require('standard-as-callback');
+var ClusterSubscriber = require('./ClusterSubscriber').default;
+var DelayQueue = require('./DelayQueue').default;
+var PromiseContainer = require('../promiseContainer');
+var {AbortError} = require('redis-errors');
+
+/**
+ * Creates a Redis Cluster instance
+ *
+ * @constructor
```
Comment (SimonSchick) [inline lib/cluster/index.ts]: You can probably omit most of those `@constructor` and other class related hints.

File: lib/cluster/index.ts [hunk 2]
```diff
@@ -0,0 +1,663 @@
+import ClusterAllFailedError from '../errors/ClusterAllFailedError'
+import {defaults, noop} from '../utils/lodash'
+import ConnectionPool from './ConnectionPool'
+import {NodeKey, IRedisOptions, normalizeNodeOptions} from './util'
+
+var Deque = require('denque')
```
Comment (SimonSchick) [inline lib/cluster/index.ts]: Some of those requires can probably be moved to import (namely node and some of your own).
Everything should probably be `const`

File: lib/cluster/index.ts [hunk 3]
```diff
@@ -0,0 +1,663 @@
...
+      this.subscriber.start();
+    });
+  };
+
+  /**
+   * Called when closed to check whether a reconnection should be made
+   *
+   * @private
+   */
+  _handleCloseEvent () {
```
Comment (SimonSchick) [inline lib/cluster/index.ts]: You can get rid of `@public` and `@private` and use the keywords instead.

File: lib/cluster/index.ts [hunk 4]
```diff
@@ -0,0 +1,646 @@
...
+      setImmediate(function () {
+        this.setStatus('close')
+        this.handleCloseEvent()
+      }.bind(this))
+
+      return ret
+    }
+    return asCallback(
+      Promise.all(this.nodes().map(function (node) {
+        return node.quit()
```
Comment (AVVS) [inline lib/cluster/index.ts]: this sometimes would throw if node that we issue quit to is already in the disconnected state (often in the tests where redis instance is down before the quit is processed)

File: lib/cluster/index.ts [hunk 5]
```diff
@@ -0,0 +1,646 @@
...
+   * Change cluster instance's status
+   *
+   * @private
+   * @param {ClusterStatus} status
+   * @memberof Cluster
+   */
+  private setStatus(status: ClusterStatus): void {
+    debug('status: %s -> %s', this.status || '[empty]', status)
+    this.status = status
+    process.nextTick(this.emit.bind(this, status))
```
Comment (AVVS) [inline lib/cluster/index.ts]: this.emit.bind could be saved as a reference, ie boundEmit
and then we can `process.nextTick(this.boundEmit, status)` - that should be a bit faster

File: lib/cluster/index.ts (hunk 4 above)
Comment (luin) [reply lib/cluster/index.ts]: Good catch! I'll fix this in a separate commit.

**Pull request merged**